tree/treeplayer/src/TTreeReaderArray.cxx

```diff
@@ -532,7 +532,25 @@ void ROOT::Internal::TTreeReaderArrayBase::CreateProxy()
          return;
       }
 
-      if (fDict != branchActualType) {
+      auto matchingDataType = [](TDictionary *left, TDictionary *right) -> bool {
+         if (left != right) {
+            auto left_datatype = dynamic_cast<TDataType *>(left);
+            auto right_datatype = dynamic_cast<TDataType *>(right);
+            if (left_datatype && right_datatype) {
+               if (left_datatype->GetType() > 0 && left_datatype->GetType() == right_datatype->GetType())
+                  return true;
+               else
+                  return ((left_datatype->GetType() == kDouble32_t && right_datatype->GetType() == kDouble_t)
```

Comment:
Consider std::sort()ing the two types so only one comparison is ever needed, not both directions.

Reply:
Do you mean:
```C++
        else
            return (  (l == kDouble32_t && r == kDouble_t)
                   || (l == kDouble_t && r == kDouble32_t)
                   || (l == kFloat16_t && r == kFloat_t)
                   || (l == kFloat_t && r == kFloat16_t));
```
vs (the slightly more fragile since it relies on the numerical order of the enums):
```c++
         else {
            if (l > r)
               std::swap(l,r);
            return (  (l == kDouble_t && r == kDouble32_t)
                   || (l == kFloat_t && r == kFloat16_t));
         }
```
?